### PR TITLE
Register through API

### PR DESCRIPTION
**Pull request by tcitworld**
Signed-off-by: Thomas Citharel <tcit@tcit.fr>

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Documentation | no
| Translation   | no
| Fixed tickets | #3056 
| License       | MIT

### CODE REVIEW

File: src/Wallabag/ApiBundle/Controller/UserRestController.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            $json = $this->get('serializer')->serialize(['error' => "Server doesn't allow registrations"], 'json');
+            return (new JsonResponse())->setJson($json)->setStatusCode(403);
+        }
+
+        if ($password === '') { // TODO : might be a good idea to enforce restrictions here
+            $json = $this->get('serializer')->serialize(['error' => 'Password is blank'], 'json');
+            return (new JsonResponse())->setJson($json)->setStatusCode(400);
+        }
+
+
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/UserRestController.php]: Empty line

File: src/Wallabag/ApiBundle/Controller/UserRestController.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            return (new JsonResponse())->setJson($json)->setStatusCode(409);
+        }
+
+        $em = $this->get('doctrine.orm.entity_manager');
+
+        $userManager = $this->get('fos_user.user_manager');
+        $user = $userManager->createUser();
+
+        $user->setUsername($username);
+
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/UserRestController.php]: You can remove all these empty lines

File: src/Wallabag/ApiBundle/Controller/UserRestController.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            return (new JsonResponse())->setJson($json)->setStatusCode(403);
+        }
+
+        if ($password === '') { // TODO : might be a good idea to enforce restrictions here
+            $json = $this->get('serializer')->serialize(['error' => 'Password is blank'], 'json');
+            return (new JsonResponse())->setJson($json)->setStatusCode(400);
+        }
+
+
+        // TODO : Make only one call to database by using a custom repository method
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/UserRestController.php]: 👍

File: src/Wallabag/ApiBundle/Controller/UserRestController.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        $event = new UserEvent($user);
+        $this->get('event_dispatcher')->dispatch(FOSUserEvents::USER_CREATED, $event);
+
+        $serializationContext = SerializationContext::create()->setGroups(['user_api']);
+        $json = $this->get('serializer')->serialize($user, 'json', $serializationContext);
+
+        return (new JsonResponse())->setJson($json);
+
+    }
+
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/UserRestController.php]: Empty lines to remove too

**Comment by j0k3r** [comment]
The registration is now working.
I re-used the form for the registration to re-use validation rules.
If an error occurs, server will respond a 400 with content:

```json
{
    "error": {
        "username": [
            "This value is already used."
        ]
    }
}
```

When a user is created, information will be send like:

```json
{
    "id": 4,
    "username": "google",
    "email": "wallabag@google.com",
    "created_at": "2017-05-30T10:42:22+0200",
    "updated_at": "2017-05-30T10:42:22+0200"
}
```

File: tests/Wallabag/ApiBundle/Controller/UserRestControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        $this->assertEquals(400, $this->client->getResponse()->getStatusCode());
+
+        $content = json_decode($this->client->getResponse()->getContent(), true);
+
+        $this->assertArrayHasKey('error', $content);
+        $this->assertArrayHasKey('username', $content['error']);
+        $this->assertArrayHasKey('email', $content['error']);
+
+        // $this->assertEquals('fos_user.username.already_used', $content['error']['username'][0]);
```
Comment (nicosomb) [inline tests/Wallabag/ApiBundle/Controller/UserRestControllerTest.php]: We can remove these comments.

File: src/Wallabag/ApiBundle/Controller/UserRestController.php [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        $user = $userManager->createUser();
+        // enable created user by default
+        $user->setEnabled(true);
+
+        $form = $this->createForm('Wallabag\UserBundle\Form\NewUserType', $user, [
+            'csrf_protection' => false,
+        ]);
+
+        // simulate form submission
+        $form->submit([
```
Comment (nicosomb) [inline src/Wallabag/ApiBundle/Controller/UserRestController.php]: Nice part 👍

File: tests/Wallabag/ApiBundle/Controller/UserRestControllerTest.php (hunk 5 above)
Comment (j0k3r) [inline tests/Wallabag/ApiBundle/Controller/UserRestControllerTest.php]: I don't know.
It's still a bit mysterious to me that text is actually in english when the translator is disabled. It should be `fos_user.username.already_used` but I don't know why it's actually translated. :confused:

**Pull request merged**